File: src/base/go2_driver_py/go2_driver_py/driver.py
"""
  需求：
      1.发布里程计消息；
      2.广播里程计坐标变换；
      3.发布关节状态消息。
  分析1：
      1.了解里程计消息字段；
      2.从哪获取？机器人已经发布高层消息话题；
      3.实现，先订阅状态话题，然后转换成里程计消息，最后发布。
  分析2：
      1.需要发布机器人基坐标系与odom坐标系的相对关系；
      2.这些相对坐标系与里程计消息类似
      3.发布。
  分析3：
      1.了解关节状态信息；
      2.获取数据
      3.订阅低层状态话题，然后解析转换成关节消息，最后发布。
"""

import rclpy
from rclpy.node import Node
from nav_msgs.msg import Odometry
from unitree_go.msg import SportModeState,LowState,MotorState
from geometry_msgs.msg import TransformStamped
from tf2_ros import TransformBroadcaster
from sensor_msgs.msg import JointState


class Driver(Node):
    def __init__(self):
        super().__init__('driver_py')

        self.declare_parameter("odom_frame","odom")
        self.declare_parameter("base_frame","base")
        self.declare_parameter("publish_tf","true")

        self.odom_frame = self.get_parameter("odom_frame").value
        self.base_frame = self.get_parameter("base_frame").value
        self.publish_tf = self.get_parameter("publish_tf").value

        # 创建 odom 的发布对象
        self.odom_pub = self.create_publisher(Odometry,"odom",10)
        # 创建 SportState 的订阅对象
        self.mode_sub = self.create_subscription(SportModeState,"/lf/sportmodestate",self.mode_cb,10)
        # 创建坐标变换广播器
        self.tf_bro = TransformBroadcaster(self)

        # 创建关节发布对象
        self.joint_pub = self.create_publisher(JointState,"joint_states",10)
        # 创建低层状态获取对象（回调函数中实现数据转换）
        self.state_sub = self.create_subscription(LowState,"/lf/lowstate",self.state_cb,10)

    def state_cb(self,state : LowState):
        # 获取关节状态并发布
        joint_state = JointState()
        # 数据组织
        joint_state.header.stamp = self.get_clock().now().to_msg()

        # 设置关节名称
        joint_state.name = [
            "FL_hip_joint","FL_thigh_joint","FL_calf_joint",
            "FR_hip_joint","FR_thigh_joint","FR_calf_joint",
            "RL_hip_joint","RL_thigh_joint","RL_calf_joint",
            "RR_hip_joint","RR_thigh_joint","RR_calf_joint",
        ]

        for i in range(12):
            q = float(state.motor_state[i].q)
            joint_state.position.append(q)


        # 发布
        self.joint_pub.publish(joint_state)

    def mode_cb(self,mode : SportModeState):
        #解析生成odom对象
        odom = Odometry()

        # 时间戳
        odom.header.stamp = self.get_clock().now().to_msg()
        # 原点坐标系
        odom.header.frame_id = self.odom_frame
        # 机器狗基坐标系
        odom.child_frame_id = self.base_frame

        # 位置
        odom.pose.pose.position.x = float(mode.position[0])
        odom.pose.pose.position.y = float(mode.position[1])
        odom.pose.pose.position.z = float(mode.position[2])

        # 姿态
        odom.pose.pose.orientation.w = float(mode.imu_state.quaternion[0])
        odom.pose.pose.orientation.x = float(mode.imu_state.quaternion[1])
        odom.pose.pose.orientation.y = float(mode.imu_state.quaternion[2])
        odom.pose.pose.orientation.z = float(mode.imu_state.quaternion[3])

        # 速度
        odom.twist.twist.linear.x = float(mode.velocity[0])
        odom.twist.twist.linear.y = float(mode.velocity[1])
        odom.twist.twist.linear.z = float(mode.velocity[2])

        odom.twist.twist.angular.z = float(mode.yaw_speed)

        # 发布
        self.odom_pub.publish(odom)

        # 广播坐标变换
        if self.publish_tf:
            # 生成坐标变换数据
            trans_form = TransformStamped()

            trans_form.header.stamp = self.get_clock().now().to_msg()
            trans_form.header.frame_id = self.odom_frame

            trans_form.child_frame_id = self.base_frame

            # 设置偏移量
            trans_form.transform.translation.x = odom.pose.pose.position.x
            trans_form.transform.translation.y = odom.pose.pose.position.y
            trans_form.transform.translation.z = odom.pose.pose.position.z

            # 设置旋转角度
            trans_form.transform.rotation = odom.pose.pose.orientation

            # 发布
            self.tf_bro.sendTransform(trans_form)

def main():
    rclpy.init()
    rclpy.spin(Driver())
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/base/go2_driver/src/driver.cpp
/*
  需求：
      1.发布里程计消息；
      2.广播里程计坐标变换；
      3.发布关节状态消息。
  分析1：
      1.了解里程计消息字段；
      2.从哪获取？机器人已经发布高层消息话题；
      3.实现，先订阅状态话题，然后转换成里程计消息，最后发布。
  分析2：
      1.需要发布机器人基坐标系与odom坐标系的相对关系；
      2.这些相对坐标系与里程计消息类似
      3.发布。
  分析3：
      1.了解关节状态信息；
      2.获取数据
      3.订阅低层状态话题，然后解析转换成关节消息，最后发布。

*/

#include "rclcpp/rclcpp.hpp"
#include "nav_msgs/msg/odometry.hpp"
#include "unitree_go/msg/sport_mode_state.hpp"
#include "geometry_msgs/msg/transform_stamped.hpp"
#include "tf2_ros/transform_broadcaster.h"
#include "sensor_msgs/msg/joint_state.hpp"
#include "unitree_go/msg/low_state.hpp"

using namespace std::placeholders;
class Driver : public rclcpp::Node
{
public:
    Driver() : Node("driver")
    {
        // 构造函数：初始化发布者、订阅者、服务等
        // 声明参数
        this->declare_parameter("odom_frame","odom");
        this->declare_parameter("base_frame","base");
        this->declare_parameter("publish_tf",true);

        // 获取参数
        odom_frame = this->get_parameter("odom_frame").as_string();
        base_frame = this->get_parameter("base_frame").as_string();
        publish_tf = this->get_parameter("publish_tf").as_bool();

        // 坐标变换广播器
        tf_bro_ = std::make_unique<tf2_ros::TransformBroadcaster>(*this);


        // RCLCPP_INFO(this->get_logger(), "Hello from my ROS 2 node!");

        RCLCPP_INFO(this->get_logger(), "Using frames: odom='%s', base='%s'", 
                odom_frame.c_str(), base_frame.c_str());

        odom_pub_ = this->create_publisher<nav_msgs::msg::Odometry>("odom",10);
        mode_sub_ = this->create_subscription<unitree_go::msg::SportModeState>(
                    "/lf/sportmodestate",10,std::bind(&Driver::mode_cb,this,_1));
        //关节状态发布
        joint_state_pub_ = this->create_publisher<sensor_msgs::msg::JointState>("joint_states",10);
        low_state_sub_ = this->create_subscription<unitree_go::msg::LowState>(
          "lf/lowstate",10,std::bind(&Driver::low_state_cb,this,_1)
        );
    }

private:
    // 发布关节信息
    rclcpp::Publisher<sensor_msgs::msg::JointState>::SharedPtr joint_state_pub_;
    //订阅低层状态
    rclcpp::Subscription<unitree_go::msg::LowState>::SharedPtr low_state_sub_;
    //广播坐标变换
    std::unique_ptr<tf2_ros::TransformBroadcaster> tf_bro_;
    // 订阅go2状态
    rclcpp::Subscription<unitree_go::msg::SportModeState>::SharedPtr mode_sub_;
    // 发布里程计
    rclcpp::Publisher<nav_msgs::msg::Odometry>::SharedPtr odom_pub_;
    std::string odom_frame,base_frame;
    bool publish_tf;
    // 订阅低层信息获取关节状态，组织消息并发布
    void low_state_cb(const unitree_go::msg::LowState::SharedPtr low_state){
      sensor_msgs::msg::JointState joint_state;

      //组织数据
      joint_state.header.stamp = this->now();
      // 描述关节名称
      joint_state.name = {
        "FL_hip_joint","FL_thigh_joint","FL_calf_joint",
        "FR_hip_joint","FR_thigh_joint","FR_calf_joint",
        "RL_hip_joint","RL_thigh_joint","RL_calf_joint",
        "RR_hip_joint","RR_thigh_joint","RR_calf_joint",
      };

      // 遍历低层状态信息的关节数据
      for (size_t i = 0; i < 12; i++)
      {
        auto motor = low_state->motor_state[i];// 获取某个关节信息
        // 获取旋转角度，添加进
        joint_state.position.push_back(motor.q);
      }

      joint_state_pub_->publish(joint_state);
    }

    // 订阅
    void mode_cb(const unitree_go::msg::SportModeState::SharedPtr mode){
      // 转换并发布里程计
      nav_msgs::msg::Odometry odom;
      //组织数据
      //头
      odom.header.stamp.sec = mode->stamp.sec;
      odom.header.stamp.nanosec = mode->stamp.nanosec;
      odom.header.frame_id = "odom_frame"; // 里程计坐标系
      odom.child_frame_id = "base_frame"; //机器人坐标系 
      //位姿
      odom.pose.pose.position.x = mode->position[0];
      odom.pose.pose.position.y = mode->position[1];
      odom.pose.pose.position.z = mode->position[2];

      odom.pose.pose.orientation.w = mode->imu_state.quaternion[0];
      odom.pose.pose.orientation.x = mode->imu_state.quaternion[1];
      odom.pose.pose.orientation.y = mode->imu_state.quaternion[2];
      odom.pose.pose.orientation.z = mode->imu_state.quaternion[3];
      //速度
      odom.twist.twist.linear.x = mode->velocity[0];
      odom.twist.twist.linear.y = mode->velocity[1];
      odom.twist.twist.linear.z = mode->velocity[2];

      odom.twist.twist.angular.z = mode->yaw_speed;

      odom_pub_->publish(odom);

      if (!publish_tf)
      {
        return;
      }

      // 发布坐标变换
      geometry_msgs::msg::TransformStamped transform;
      transform.header.stamp = this->now();
      transform.header.frame_id = odom_frame;
      transform.child_frame_id = base_frame;
      
      //设置偏移量
      transform.transform.translation.x = odom.pose.pose.position.x;
      transform.transform.translation.y = odom.pose.pose.position.y;
      transform.transform.translation.z = odom.pose.pose.position.z;

      //设置偏移角度
      transform.transform.rotation = odom.pose.pose.orientation;

      tf_bro_->sendTransform(transform);
    }

};

int main(int argc, char ** argv)
{
    // 初始化 ROS 2
    rclcpp::init(argc, argv);
    // 创建节点并开始自旋（处理回调）
    auto node = std::make_shared<Driver>();
    rclcpp::spin(node);
    // 关闭 ROS 2
    rclcpp::shutdown();
    return 0;
}